src/interop.ts

```diff
@@ -12,25 +12,304 @@
+        }
+    }
+    return { nativeType: NativeType.NON_JSON };
+}
+
+/**
+ * Determines whether or not the property in question is a JSON type of not.
+ * If the property is a nested property then the `propName` will contain all the parent
+ * properties so that the correct nested type can be found
+ *
+ * @param properties The resource properties
+ * @param propName the property name as a list containing parent property names
+ * @param types The pulumi types
+ * @returns true if the property is a JSON type and should not be normalized
+ */
+function isJsonType(
+    propName: string[],
+    properties: { [key: string]: PulumiProperty },
+    types: { [key: string]: PulumiType },
+): boolean {
+    let props = properties;
+    for (let i = 0; i < propName.length; i++) {
+        const prop = toSdkName(propName[i]);
+        if (prop in props) {
+            const metaProp = props[prop];
+            const { nativeType, meta } = processMetadataProperty(i + 1 === propName.length, metaProp, types);
+            if (nativeType === NativeType.JSON) {
+                return true;
+            }
+            props = meta!;
+        }
+    }
+    return false;
+}
+
+/**
+ * The pulumi provider to read the schema from
+ */
+export enum PulumiProvider {
+    // We currently only support aws-native provider resources
+    AWS_NATIVE = 'AWS_NATIVE',
+}
+
+/**
+ * Recursively normalizes object types, with special handling for JSON types (which should not be normalized)
+ *
+ * @param key the property key as a list (including parent property names for nested properties)
+ * @param value the value to normalize
+ * @param cfnType The CloudFormation resource type being normalized (e.g. AWS::S3::Bucket). If no value
+ * is provided then property conversion will be done without schema knowledge
+ * @param pulumiProvider The pulumi provider to read the schema from. If `cfnType` is provided then this defaults
+ * to PulumiProvider.AWS_NATIVE
+ * @returns the normalized property value
+ */
+function normalizeObject(key: string[], value: any, cfnType?: string, pulumiProvider?: PulumiProvider): any {
+    if (!value) return value;
+    if (Array.isArray(value)) {
+        const result: any[] = [];
+        for (let i = 0; i < value.length; i++) {
+            result[i] = normalizeObject(key, value[i], cfnType);
+        }
+        return result;
+    }
+
+    if (typeof value !== 'object' || pulumi.Output.isInstance(value) || value instanceof Promise) {
+        return value;
+    }
+
+    const result: any = {};
+    if (cfnType) {
+        try {
+            const metadata = Metadata.instance(pulumiProvider ?? PulumiProvider.AWS_NATIVE);
+            const resource = metadata.findResource(cfnType);
+            if (isJsonType(key, resource.inputs, metadata.types())) {
+                return value;
+            }
+
+            Object.entries(value).forEach(([k, v]) => {
+                result[toSdkName(k)] = normalizeObject([...key, k], v, cfnType);
```

Comment:
I'm thinking toSdkName is probably approximate, but still not 100% sure. Some corner cases likely lurking. I've been reading this code recently:

https://github.com/pulumi/pulumi-aws-native/blob/master/provider/pkg/naming/naming.go#L24

And ToCfnName seems to need a table. But ToSdkName doesn't?  That's very interesting.  I suspect as the system grows both will need to consult tables from the native metadata. We can leave for later though.

Reply:
I was looking into the `aws-native` implementation as well. I'll add a backlog item for making sure the the `pulumi-cdk` renaming libraries match `aws-native`.

Reply:
Actually I think  I can just handle this in this PR. I forgot that I already did the conversion in an earlier PR, but without the `lookupTable` because we didn't have the metadata which we now do!

Reply:
nvm I got confused. `toSdkName` doesn't require a lookup table and it is the only one currently in use in this library.